### PR TITLE
feat: select caller hosts during setup

### PR DESCRIPTION
**Pull request by piwi3910**
## What

Make init and agents select explicit or reliably identified caller hosts, print additive host declarations and rule files, and validate declared or existing integrations without requiring every supported host.

## Why

A single-host repository should not acquire unrelated integrations merely because it adopts one rule file. Unknown caller context requires an explicit choice. Missing shared contracts must not allow partial setup or silently disable existing integrations.

## How

Strict repeatable --host flags or exclusive --all select targets. Caller context travels through CLI/API environments and Kilo/OpenCode/pi adapters. .procoder/hosts.json records additive choices; generated content remains print-only. Init preserves existing ignores while adding selected host paths. Shared-contract validation precedes generated output, ignore writes and tool installs. Canonical and host command examples preserve user arguments on initial and repeat invocations. This distribution explicitly opts into all hosts.

## Testing

procoder test passed 57 Go packages on the final content. Standalone JavaScript NOT run: no test script; Node-driven adapter tests run through Go. Focused packages passed uncached. The full 33-file range gate passed with zero unformatted, unchecked or blocking findings. git diff --check passed. Mutation experiments produced observed failures, were restored and verified, and now have proved-by markers and todo evidence. Adversarial and edge-case review passes completed. Final CI https://github.com/azrtydxb/procoder/actions/runs/35274890757 passed gate and Ubuntu/macOS/Windows tests; CodeQL checks passed.

## Notes for the reviewer

Dependency #285 merged as 30f472551d24525f0f24149223473ec851ace0b3. This PR targets main and incorporates it through normal merge eddc19d, without rewriting history. Dependency files match main exactly and are absent from the feature diff. Precondition fixes: ce606ef. Argument forwarding and observed mutation evidence: 4f262b1. All eight Copilot review threads have evidence replies and are resolved. The preexisting release-date decision from 92f7b02 is preserved byte-for-byte and remains unanswered. No release or tag changes are included.

### CODE REVIEW

**Comment by piwi3910** [comment]
Verification completed on head 92f7b023761d852a218327eacc66ab52bc39c7c4: https://github.com/azrtydxb/procoder/actions/runs/35269024892 concluded SUCCESS. Gate and all three OS test jobs passed; docs and release jobs were skipped. Local procoder test passed 57 Go packages; the explicit 34-file branch-range gate had zero blocking findings. Spec and plan checks both reported COMPLETE.

Merge remains intentionally blocked on #285, whose two review conversations are still unresolved. This draft is based on that dependency branch to isolate the host-aware changes; it must not merge into that branch. Resolve and merge #285, retarget this PR to main, complete feature review, and reverify the resulting head before merging. No release-date answer, release/tag operation, branch deletion, or force operation was performed.

**Comment by piwi3910** [comment]
Feature defect pass fixed two valid issues in ce606ef: missing/unreadable AGENTS.md previously emitted a declaration and let init change .gitignore; existing host copies without a declaration silently passed when their master was missing. TestSetupRequiresReadableMasterBeforeOutputOrWrites and TestExistingHostWithoutDeclarationStillRequiresMaster both failed before the fix and pass now. Full procoder test passed 57 Go packages after integrating main in eddc19d; full 33-file branch-range gate has zero blocking findings. Dependency conflict hunks were resolved to the reviewed #285 result and verified byte-identical to main; feature changes and the pending release-date decision were preserved. Awaiting new required CI on the retargeted PR.

File: commands/agents.md [hunk 1]
```diff
@@ -4,6 +4,12 @@ description: "Keep the universal agent layer in sync: per-host rule files derive
 
 The user invoked /procoder:agents.
 
+Pass any user-selected `--host <name>` flags or exclusive `--all` to the
+command. Without flags it uses reliable plugin context; if unknown, ask the
+user which host and re-run with that explicit choice. Never choose all by
+default. Review and write the printed `.procoder/hosts.json` declaration along
+with the named copies. Additional hosts are additive; keep existing files.
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply commands/agents.md:11]: Fixed in 4f262b1: both agents invocations now retain $ARGUMENTS, as do the OpenCode and Kilo twins. TestSetupCommandExamplesForwardArguments failed for all six command files before the fix and passes now. Command parity tests and the full 57-package Go suite pass.

File: internal/portability/opencode_test.go [hunk 2]
```diff
@@ -112,3 +112,30 @@ func TestOpenCodeTurnEndRunsHookStop(t *testing.T) {
 		}
 	}
 }
+
+func TestOpenCodeAndKiloSupplyExplicitSetupContext(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply internal/portability/opencode_test.go:118]: Addressed in 4f262b1. Mutated shell.env to assign outer rather than the host; ran TestOpenCodeAndKiloSupplyExplicitSetupContext with -count=1 and observed both Node subtests fail with PROCODER_HOST=outer. Added the observed proved-by marker, restored the adapter, verified no remaining production diff, and reran focused packages and the full suite successfully.

File: cmd/procoder/setup_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"procoder/internal/api"
+	"procoder/internal/host"
+)
+
+func TestSetupCLIAPIAndCallerIsolation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply cmd/procoder/setup_test.go:16]: Addressed in 4f262b1 with observed proved-by markers on all three tests. Unconditional all-host Setup made the no-context CLI/API init assertion fail; no-op IgnoreHosts failed the selected-ignore assertion; the master-precondition test was observed failing before ce606ef (declaration output and ignore writes after missing AGENTS.md). Mutations were restored with patches and verified by git diff --exit-code. Uncached focused packages and procoder test (57 Go packages) pass. Full experiment evidence is in .procoder/todo/20260917-fix-host-setup-precondition-failures.md.

File: commands/init.md [hunk 4]
```diff
@@ -4,12 +4,24 @@ description: Install the formatters this repository needs, with every command vi
 
 The user invoked /procoder:init.
 
+Use the user's explicit `--host <name>` (repeatable for additional hosts) or
+`--all` selection when provided. Otherwise the binary uses reliable plugin
+context. If it asks which host to set up, ask the user and re-run with their
+choice; never assume all hosts or infer one from repository directories.
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply commands/init.md:10]: Fixed in 4f262b1: initial and repeat init invocations now explicitly retain $ARGUMENTS, including both host twins. TestSetupCommandExamplesForwardArguments failed before the fix and passes now; command parity tests also pass. procoder test passed 57 Go packages and procoder check passed with zero blocking findings.

File: internal/portability/portability_test.go [hunk 5]
```diff
@@ -49,8 +49,7 @@ func TestEveryDeclaredCopyShips(t *testing.T) {
 }
 
 // A repo with AGENTS.md but no copies has not adopted the layer — silence.
-// Once one copy exists (adoption), drift blocks and the other copies are
-// reported missing as information.
+// Once one copy exists, drift blocks without demanding unrelated hosts.
 func TestDriftBlocksAndMissingInformsOnceAdopted(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply internal/portability/portability_test.go:55]: Addressed in 4f262b1. Forced selectedCopies to all and ran TestDriftBlocksAndMissingInformsOnceAdopted with -count=1; it failed on twelve missing integrations in an undeclared layer. Added the observed proved-by marker. Also exercised both modified drift tests and corrected the stale unreadable-copy mutation comment. Restored production code; focused uncached tests and full procoder test pass.

File: internal/host/setup_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
+package host
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestSetupSelection(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply internal/host/setup_test.go:10]: Addressed in 4f262b1. Applied and ran two mutations: unconditional all-host Setup caused TestSetupSelection to reject accepted invalid/unknown requests and wrong explicit selection; removing PROCODER_HOST from envKeys caused TestSetupContextTravelsInProcessEnvironment to fail on lost caller context. Both tests now have observed proved-by comments. Mutations restored, focused uncached packages and full procoder test pass.

File: internal/portability/selection_test.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"procoder/internal/gitx"
+)
+
+func TestHostSetupPrintsOnlyNeededFilesWithoutWriting(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply internal/portability/selection_test.go:14]: Addressed in 4f262b1. Forced all-copy selection and observed the print-only and declared/existing tests fail; returning no declaration without reading caused missing-master and dangling-declaration tests to fail with empty findings. The installed-copy missing-master regression was observed failing before ce606ef. The new command-forwarding test failed for all six command files before the fix. Every test now names its observed failure; mutations restored and checked with git diff --exit-code. Focused uncached packages, full procoder test (57 Go packages), and procoder check pass. Detailed evidence is in the feature todo.

File: internal/initcmd/gitignore_test.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import (
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func TestIgnoreHostsPreservesExistingRulesAndIsIdempotent(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (piwi3910) [reply internal/initcmd/gitignore_test.go:14]: Addressed in 4f262b1. Applied a successful early return in IgnoreHosts and ran all five gitignore tests with -count=1: missing .gitignore, failed git check-ignore results, accepted directory and symlink inputs, and absent selected entry all produced failures. Each test now names its observed failure in a proved-by comment. Restored production code and verified restoration with git diff --exit-code; focused uncached packages and full procoder test pass. Experiment evidence is recorded in the feature todo.

**Comment by piwi3910** [comment]
All eight review findings are addressed in 4f262b1 with evidence replies. Setup arguments now survive initial and repeat commands across canonical/OpenCode/Kilo files. Mutation experiments were actually applied and targeted tests were observed failing, then production code was restored and verified unchanged. New and modified setup tests have observed proved-by markers; the feature todo records the experiments. Focused packages passed uncached; procoder test passed 57 Go packages (standalone JS still NOT run: no script); procoder check passed 14 changed files with zero blocking findings. New CI is running on the final follow-up head.
